### ISSUE TITLE
Track Java/JDBC and go-sql-spanner feature parity

### ISSUE THREAD

**Issue by apstndb** [apstndb/spanner-mycli#47]
# Java/JDBC and go-sql-spanner feature parity

This issue is the implementation roadmap for CLI-relevant driver features. The source compatibility matrix remains [docs/spanner-driver-compatibility.md](https://github.com/apstndb/spanner-mycli/blob/main/docs/spanner-driver-compatibility.md); update it through the implementation PRs. Its old “candidate gaps (no issue yet)” entries now have the tickets below. Do not duplicate those candidates or revive obsolete session-pool settings.

## Audit baseline (2026-09-14)

- spanner-mycli: `a8fa2418b65af9b889cf92e02190366b3e6e02ea`, released as v0.35.0-alpha.2.
- Current Java Connection API/JDBC sources: [google-cloud-java `dd2c41a`](https://github.com/googleapis/google-cloud-java/tree/dd2c41a0e2ec05d39b93f7d94169d596bb7a5483/java-spanner) and its java-spanner-jdbc subtree. The separate java-spanner and java-spanner-jdbc repositories are archived.
- Go SQL driver: [`28b0df2`](https://github.com/googleapis/go-sql-spanner/blob/28b0df2646750c7aab56da320fe6fc945b30ec86/connection_properties.go).
- [Official JDBC client statements](https://docs.cloud.google.com/spanner/docs/jdbc-session-mgmt-commands) supplement source evidence. Java PostgreSQL syntax is not automatically a GoogleSQL syntax guarantee.

Source-backed inspection covered statement parsers/registries, transaction and DDL execution, configuration and current Go SDK support. The named-parameter case mismatch in #958 was also reproduced with the actual alpha binary and a local emulator. Other gaps are source-backed, not claims of live managed-Spanner verification.

## First implementation wave

Keep one independently reviewable behavior per PR.

- [ ] #957 — MAX_PARTITIONS server hint, independent of worker parallelism.
- [ ] #958 — case-insensitive named parameter binding.
- [ ] #959 — SHOW TRANSACTION isolation/read-only inspection.
- [ ] #482 — logical transaction deadline, preserving one budget across replay.
- [ ] #484 — shared startup-snapshot reset primitive and RESET ALL.
- [ ] #960 — single-variable RESET, blocked by #484.

## Second wave

- [ ] #961 — separate COMMIT_PRIORITY.
- [ ] #962 — full DirectedReadOptions JSON, preserving the existing shorthand.
- [ ] #965 — configurable keepalive, preserving owner/attempt isolation.
- [ ] #485 — DDL_EXECUTION_MODE and bounded ASYNC_WAIT.
- [ ] #357 — user-idle transaction expiry; coordinate with #965 and #482. Its primary parent remains #342.
- [ ] #83 — AUTOCOMMIT using the pending/logical-owner lifecycle.
- [ ] #45 — cross-process RUN PARTITION, including token contents and resource lifetime.

The reset, timeout, AUTOCOMMIT and partition-token tickets carry explicit lifecycle conditions. A small surface area does not remove the need to settle those conditions before coding.

## Design before implementation

- [ ] #293 — bounded ABORTED retry. Reuse SAVEPOINT capture/fingerprints; do not build another journal. Agree on default/overhead and the #482 budget contract before rollout.
- [ ] #401 — expected automatic-batch counts and verification; do not present buffered counts as actual affected rows.
- [ ] #402 — DDL inside transactions, including successful commit followed by failed DDL.
- [ ] #963 — explicit sequence-kind repair; this changes a database-wide option and must preserve partially applied DDL.
- [ ] #964 — mutation-limit-triggered PDML fallback; explicitly weaker atomicity, never generic error fallback.
- [ ] #966 — first-write transaction start; Java explicitly weakens SERIALIZABLE semantics. Not equivalent to mycli's existing lazy first-operation activation.
- [ ] #663 — native/client-exported metrics with an owned exporter lifecycle.
- [ ] #967 — opt-in client/end-to-end tracing, coordinated with metrics but independently enabled.
- [ ] #968 — custom CA/client-certificate configuration for remote Omni.

These are planned design tasks, not approval to silently enable the behavior, claim full driver equivalence or ship a stable contract.

## Already supported / intentional differences

- SAVEPOINT is implemented (#404, PRs #952–#956), opt-in with synchronous replay and documented limits.
- SET LOCAL is implemented (#400/#691); transaction owners preserve undo across physical replacement.
- Directed read is implemented (#486/#907); #962 only extends its value representation and routing choices.
- Read-lock mode, default isolation, staleness, request/transaction tags, commit statistics/delay, change-stream exclusion, automatic/manual DML batching, automatic partitioning and Data Boost already have CLI support. Do not create duplicate feature tickets for them.
- Keep PostgreSQL aliases such as transaction_isolation/default_transaction_isolation and broader PG transaction syntax in #230. The existing DEFAULT_ISOLATION_LEVEL already selects isolation; an alias is not a new execution capability.
- #403 is closed because its legacy session-pool hook is a no-op; the user-idle requirement is #357.
- JDBC ResultSet/PreparedStatement contracts, Java thread/session pools, database/sql caches and driver-only type decoding are not features to port into this CLI. Java's manual BATCH_DML_UPDATE_COUNT is an ORM provisional-count contract; reconsider only if a CLI use case emerges (#401).
- Native leader routing is already the SDK default. A separate diagnostic override, channel tuning, generic connection-state emulation and DSN-only options are deferred until a concrete CLI use case justifies them.

## Delivery rules

Refresh current HEAD before implementation. Each PR updates the corresponding compatibility-matrix row, uses the existing varDefs/transactionContext/output paths, captures validation output and exit status, and runs `make check` before push; transaction concurrency changes also run `make check-race`. The coverage floor stays 80%. Keep generated/opaque state out of scalar reset loops. Do not infer native-server guarantees from emulator-only checks.

The 2026-07-06 audit-table history remains in this issue's edit history and comments.

**Comment by apstndb** [comment]
This comment provides a comprehensive update to this umbrella issue, based on a detailed analysis of the latest `main` branch, all open issues, and all open PRs as of 2025-07-20.

The goal is to provide a single source of truth for the current compatibility status and to identify concrete next steps.

---

### Key Findings & Discrepancies with the Original Issue Body

Our analysis revealed several points that update or supersede the information currently in the issue description:

1.  **`go-sql-spanner` is a Subset:** The feature set of `go-sql-spanner` is a functional subset of `java-spanner`. Therefore, targeting `java-spanner` for 100% compatibility will inherently cover `go-sql-spanner`. The tables below have been simplified to reflect this.
2.  **Transaction-related Implementation Status:** Several core transaction features, including `AUTOCOMMIT` and `READONLY`, are more fragile than they appear. While the commands are parsed, open Issue #329 ("refactor: Improve transaction management") and Issue #180 ("Implement `AUTOCOMMIT`") indicate that the underlying execution logic is incomplete and requires significant refactoring. They should be considered **Partial** rather than fully implemented.
3.  **Limitations in Implemented Features:** Some existing features have limitations not previously tracked here. For example, `TRANSACTION_TAG` and `STATEMENT_TAG` are not applied to Partitioned DML, as noted in Issue #345.
4.  **Support for `SET LOCAL`:** `java-spanner` allows the `[LOCAL]` keyword in `SET` statements to scope a variable change to the current transaction only. This is a subtle but important compatibility feature that `spanner-mycli` currently does not support.
5.  **Untracked Features:** A number of features available in `java-spanner` are not yet tracked in this umbrella issue. These are potential candidates for new sub-tasks.

---

### Detailed Compatibility Analysis

The following tables provide an up-to-date summary of the compatibility status.

**Status Legend:**
*   ✅ **Implemented**: Fully functional, no known open issues.
*   ⚠️ **Partial**: The basic feature exists, but it is incomplete or has known bugs outlined in an open issue.
*   ⚠️ **Enhancement Proposed**: The feature is implemented, but an open issue/PR proposes improvements or extensions.
*   ❌ **Not Implemented**: The feature is missing.

#### Runtime Feature Comparison

| Variable / Command | `java-spanner` | `spanner-mycli` Status & Notes |
| :--- | :---: | :--- |
| `AUTOCOMMIT` | `read/write [LOCAL]` | `read/write`<br>⚠️ **Partial**: Execution logic is incomplete per Issue #180 and requires refactoring per Issue #329. |
| `READONLY` | `read/write [LOCAL]` | `read/write`<br>⚠️ **Partial**: Core transaction logic requires refactoring per Issue #329. |
| `TRANSACTION` | `write-only` | `write-only`<br>⚠️ **Partial**: Core transaction logic requires refactoring per Issue #329. |
| `AUTOCOMMIT_DML_MODE` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `READ_ONLY_STALENESS` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `STATEMENT_TIMEOUT` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `OPTIMIZER_VERSION` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `OPTIMIZER_STATISTICS_PACKAGE` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `RETURN_COMMIT_STATS` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `TRANSACTION_TAG` | `read/write [LOCAL]` | `read/write`<br>⚠️ **Enhancement Proposed**: Tag is not applied to Partitioned DML per Issue #345. |
| `STATEMENT_TAG` | `read/write [LOCAL]` | `read/write`<br>⚠️ **Enhancement Proposed**: Tag is not applied to Partitioned DML per Issue #345. |
| `RETRY_ABORTS_INTERNALLY` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `DATA_BOOST_ENABLED` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `AUTO_PARTITION_MODE` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `EXCLUDE_TXN_FROM_CHANGE_STREAMS` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `MAX_COMMIT_DELAY` | `read/write [LOCAL]` | `read/write`<br>✅ **Implemented** |
| `SAVEPOINT` | `command` | ❌ **Not Implemented** |
| `BATCH DDL/DML` | `command` | `command`<br>✅ **Implemented** |
| `COMMIT_TIMESTAMP` | `read-only` | `read-only`<br>✅ **Implemented** |
| `READ_TIMESTAMP` | `read-only` | `read-only`<br>✅ **Implemented** |
| `AUTO_BATCH_DML_UPDATE_COUNT` | `read/write [LOCAL]` | ❌ **Not Implemented** |
| `AUTO_BATCH_DML_UPDATE_COUNT_VERIFICATION` | `read/write [LOCAL]` | ❌ **Not Implemented** |

#### Connection-Time Only Properties

| Property | Description | `java-spanner` | `spanner-mycli` Status |
| :--- | :--- | :---: | :--- |
| `MinSessions` / `MaxSessions` | Configure session pool size. | ✅ | ❌ **Not Implemented**. See Issue #214. |
| `NumChannels` | Configure gRPC channel count. | ✅ | ❌ **Not Implemented**. See Issue #214. |
| `autoconfigEmulator` | Auto-configure for the Spanner emulator. | ✅ | ❌ **Not Implemented** |
| `keepTransactionAlive` | Periodically pings to keep transactions alive. | ✅ | ❌ **Not Implemented**. A related but different feature (`--idle-transaction-timeout`) is requested in Issue #181. |
| `ddlInTransactionMode` | How to handle DDL in a transaction. | ✅ | ❌ **Not Implemented** |
| `Inactive Transaction Action` | Warns on or closes inactive transactions to prevent session leaks. | ✅ | ❌ **Not Implemented**. Could be implemented as a session-init-only system variable (e.g., `CLI_INACTIVE_TRANSACTION_ACTION`). |

---

### Proposed Sub-tasks

Based on this analysis, the following features are not yet implemented and do not have tracking issues. It is proposed to create new issues for them and link them back to this umbrella issue.

*   [ ] Create issue for `SAVEPOINT` support.
*   [ ] Create issue for `AUTO_BATCH_DML_UPDATE_COUNT` and `AUTO_BATCH_DML_UPDATE_COUNT_VERIFICATION`.
*   [ ] Create issue for `ddlInTransactionMode` support.
*   [ ] Create issue for `autoconfigEmulator` support.
*   [ ] Create issue for `Inactive Transaction Action` (`CLI_INACTIVE_TRANSACTION_ACTION`) support.
*   [ ] Create issue for `SET LOCAL` support.

**Comment by apstndb** [comment]
The tracking table is stale in both directions and needs a re-audit pass before driving more work:
- `READ_LOCK_MODE` is marked NOT IMPLEMENTED (→ #483), but it is fully implemented (var_registry.go, transaction_manager.go) and #483 is closed.
- `RETRY_ABORTS_INTERNALLY` is marked IMPLEMENTED and `AUTOCOMMIT` read/write, but both are registered as `UnimplementedVar` (var_registry.go) with no retry logic anywhere; the real work is tracked in #293 / #83.

**Comment by apstndb** [comment]
Checklist refresh (supersedes the 2026-02-21 audit for the rows below):
- SET LOCAL / [LOCAL] syntax: now implemented (#691; #400 closed). The "not supported" note is obsolete.
- READ_LOCK_MODE: now implemented (#483 closed).
- autoconfigEmulator: now implemented (#405 closed).
- COMMIT_RESPONSE is now registry-native (#750); no special-casing remains.
- The rename items #485 (CLI_ASYNC_DDL -> DDL_EXECUTION_MODE), #486 (CLI_DIRECT_READ -> DIRECTED_READ), and #487 (CLI_PROTO_DESCRIPTOR_FILE -> PROTO_DESCRIPTORS_FILE_PATH) are unblocked by the varDef alias mechanism (#750); #486 and #484 are scheduled phases of the #725 varDef series.

Still not implemented: AUTOCOMMIT (#83), RETRY_ABORTS_INTERNALLY (#293), PROTO_DESCRIPTORS (#295), AUTO_BATCH_DML_UPDATE_COUNT (#401), ddlInTransactionMode (#402), Inactive Transaction Action (#403), SAVEPOINT (#404), TRANSACTION_TIMEOUT (#482), RESET ALL (#484).

**Comment by apstndb** [comment]
Reference driver data point (go-sql-spanner):

The official Spanner drivers-overview page lists two sibling Go/JVM reference drivers for GoogleSQL: Spanner JDBC (java-spanner) and `database/sql` (googleapis/go-sql-spanner). This umbrella currently tracks java-spanner only. Since spanner-mycli is a Go CLI, go-sql-spanner is arguably the more natural compatibility reference, and it exposes the same client-side statement surface (`SHOW [VARIABLE]`, `SET [LOCAL|SESSION] [TRANSACTION]`, `RESET`, `START BATCH DDL|DML`, `RUN BATCH`, `ABORT BATCH`, `RUN PARTITIONED QUERY`, `BEGIN/COMMIT/ROLLBACK`) with snake_case connection property names.

Proposal: track go-sql-spanner alongside java-spanner in this umbrella. Mapping of its properties/statements to open children (source: `connection_properties.go`, `parser/statements.go`; versions from `CHANGES.md`):

| Feature | go-sql-spanner name | Shipped | Child |
|---|---|---|---|
| Async DDL modes | `ddl_execution_mode` = `SYNC`/`ASYNC`/`ASYNC_WAIT` (+ `ddl_async_wait_timeout`) | v1.24.0 | #485 |
| Directed read | `directed_read` | v1.26.0 | #486 |
| Transaction timeout | `transaction_timeout` | v1.22.0 | #482 |
| Auto-batch DML | `auto_batch_dml`, `auto_batch_dml_update_count`, `auto_batch_dml_update_count_verification` | v1.11.0 | #401 |
| Internal abort retry | `retry_aborts_internally` (default true) | pre-v1.0 | #293 |
| Autocommit | no `autocommit` property; relies on `database/sql` native autocommit; `autocommit_dml_mode` = `Transactional`/`PartitionedNonAtomic` | - | #83 |
| Partitioned query stmt | `RUN PARTITIONED QUERY <select>` (auto-partition + merge) | v1.24.0 | #45 |

Note: go-sql-spanner uses `isolation_level` (not `DEFAULT_ISOLATION_LEVEL`) and `read_lock_mode`, matching existing spanner-mycli vars closely.

**Comment by apstndb** [comment]
Reference driver data point (java-spanner / java-spanner-jdbc), complementing the go-sql-spanner mapping above:

Maintenance note first: java-spanner and java-spanner-jdbc have moved into the googleapis/google-cloud-java monorepo. Releases are now batched there (latest java-spanner 6.116.1, 2026-04-21; java-spanner-jdbc 2.38.0, 2026-04-13) and the changelogs live at google-cloud-java/java-spanner/CHANGELOG.md and google-cloud-java/java-spanner-jdbc/CHANGELOG.md. Links into the old standalone repos in this issue's history may go stale.

Connection API features shipped since this issue's 2026-02-21 audit window (or shortly before) that map to open children:

- transaction_timeout connection property: java-spanner 6.101.0 (2025-09-26) -> #482. Both reference drivers now have it (go-sql-spanner v1.22.0).
- statement_timeout accepted in the connection URL: 6.102.0 (2025-10-08). spanner-mycli already has STATEMENT_TIMEOUT.
- read_lock_mode for connections: 6.100.0 (2025-09-11). Already implemented here (#483).
- default_isolation_level property (6.90.0, 2025-03-31) plus per-transaction isolation and PostgreSQL isolation statements. Already implemented here as DEFAULT_ISOLATION_LEVEL.
- SHOW DEFAULT_TRANSACTION_ISOLATION for PG databases: 6.106.0 (2026-01-07). No spanner-mycli issue tracks the SHOW TRANSACTION statement family (also in go-sql-spanner v1.26.0) - candidate gap.
- default_sequence_kind auto-set: JDBC/PGAdapter behavior in 6.88.0 (2025-02-27), extended for CREATE SEQUENCE in 6.102.0. Converges with go-sql-spanner v1.26.0's default_sequence_kind property - candidate gap.
- SET LOCAL retry_aborts_internally: 6.83.0 (2024-12-13) -> #293 (interacts with the SET LOCAL support already shipped here).

Recent java-spanner core work is otherwise client-internal (multiplexed sessions, routing, gRPC metrics), plus one new feature area worth watching: Send/Ack Mutation Support for "Cloud Spanner Queue" (6.109.0, 2026-02-02).

**Comment by apstndb** [comment]
Slimmed this issue body: the full compatibility matrix and delta tracking now live in [docs/spanner-driver-compatibility.md](https://github.com/apstndb/spanner-mycli/blob/main/docs/spanner-driver-compatibility.md) (added in #768). This issue stays as the work-tracking umbrella; sub-issues continue to track individual features. The previous audit table remains available in this issue's edit history.
